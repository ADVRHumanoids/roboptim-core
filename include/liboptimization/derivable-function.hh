// Copyright (C) 2009 by Thomas Moulard, FIXME.
//
// This file is part of the liboptimization.
//
// liboptimization is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// liboptimization is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with liboptimization.  If not, see <http://www.gnu.org/licenses/>.

/**
 * \brief Declaration of the DerivableFunction class.
 */

#ifndef OPTIMIZATION_DERIVABLE_FUNCTION_HH
# define OPTIMIZATION_DERIVABLE_FUNCTION_HH
# include <limits>

# include <liboptimization/function.hh>

namespace optimization
{
  /// Define a derivable function.
  class DerivableFunction : public Function
  {
  public:
    /// Gradient type.
    typedef vector_t gradient_t;

    DerivableFunction (size_type n, value_type infinity =
                       std::numeric_limits<value_type>::infinity ()) throw ();

    /// Gradient.
    virtual gradient_t gradient (const vector_t&) const throw () = 0;
  };

}; // end of namespace optimization

#endif //! OPTIMIZATION_DERIVABLE_FUNCTION_HH
