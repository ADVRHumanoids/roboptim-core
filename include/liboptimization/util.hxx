// Copyright (C) 2009 by Thomas Moulard, FIXME.
//
// This file is part of the liboptimization.
//
// liboptimization is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// liboptimization is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with liboptimization.  If not, see <http://www.gnu.org/licenses/>.


/**
 * \brief Useful procedures (templated) implementation.
 */
#ifndef OPTIMIZATION_UTIL_HXX
# define OPTIMIZATION_UTIL_HXX
# include <boost/variant/apply_visitor.hpp>

# include <liboptimization/function.hh>

namespace optimization
{
  namespace detail
  {
    template <typename T>
    void
    jacobian_from_gradients (DerivableFunction::matrix_t& jac,
                             const std::vector<const T*>& c,
                             const DerivableFunction::vector_t& x)
    {
      for (unsigned i = 0; i < jac.size1 (); ++i)
        {
          DerivableFunction::jacobian_t grad = c[i]->jacobian (x);
          for (unsigned j = 0; j < jac.size2 (); ++j)
            jac (i, j) = grad(0, j);
        }
    }


    struct GenVariantPrintVisitor : public boost::static_visitor<void>
    {
      GenVariantPrintVisitor (std::ostream& o)
        : o_ (o)
      {}

      template <typename T>
      void operator () (const T& value)
      {
        o_ << value;
      }

      template <typename T>
      void operator () (const T* value)
      {
        if (!value)
          o_ << "Null pointer";
        else
          o_ << *value;
      }

    private:
      std::ostream& o_;
    };

  }; // end of namespace detail.

  template <typename T1, typename T2>
  std::ostream& operator<< (std::ostream& o, const boost::variant<T1, T2>& var)
  {
    detail::GenVariantPrintVisitor pv (o);
    boost::apply_visitor (pv, var);
    return o;
  }

  template <typename T>
  std::ostream& operator<< (std::ostream& o, const std::vector<T>& vect)
  {
    typedef typename std::vector<T>::const_iterator citer_t;

    if (vect.empty ())
        return o << "Empty vector";

    citer_t it = vect.begin ();
    o << *it;
    ++it;

    for (; it != vect.end (); ++it)
      o << ", " << *it;
    return o;
  }

  template <typename T1, typename T2>
  std::ostream& operator<< (std::ostream& o, const std::pair<T1, T2>& p)
  {
    return o << "(" << p.first << ", " << p.second << ")";
  }

}; // end of namespace optimization.

#endif //! OPTIMIZATION_UTIL_HXX
