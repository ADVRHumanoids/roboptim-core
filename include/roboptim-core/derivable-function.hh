// Copyright (C) 2009 by Thomas Moulard, AIST, CNRS, INRIA.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

/**
 * \brief Declaration of the DerivableFunction class.
 */

#ifndef ROBOPTIM_CORE_DERIVABLE_FUNCTION_HH
# define ROBOPTIM_CORE_DERIVABLE_FUNCTION_HH
# include <limits>

# include <roboptim-core/function.hh>

namespace roboptim
{
  /// Define a derivable function.
  class DerivableFunction : public Function
  {
  public:
    /// Gradient type.
    typedef vector_t gradient_t;
    /// Jacobian type.
    typedef matrix_t jacobian_t;

    DerivableFunction (size_type n, size_type m = 1) throw ();

    /// Jacobian.
    virtual jacobian_t jacobian (const vector_t&) const throw ();

    /// Gradient.
    virtual gradient_t gradient (const vector_t&, int) const throw () = 0;

    virtual std::ostream& print (std::ostream&) const throw ();
  };

} // end of namespace roboptim

#endif //! ROBOPTIM_CORE_DERIVABLE_FUNCTION_HH
