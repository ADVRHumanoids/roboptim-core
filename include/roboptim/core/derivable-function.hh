// Copyright (C) 2009 by Thomas Moulard, AIST, CNRS, INRIA.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

#ifndef ROBOPTIM_CORE_DERIVABLE_FUNCTION_HH
# define ROBOPTIM_CORE_DERIVABLE_FUNCTION_HH
# include <limits>
# include <utility>

# include <roboptim/core/function.hh>

namespace roboptim
{
  /// \addtogroup roboptim_meta_function
  /// @{

  /// \brief Define an abstract derivable function (\f$C^1\f$).
  ///
  /// A derivable function which provides a way to compute its
  /// gradient/jacobian.
  ///
  /// \f[ f : x \rightarrow f(x) \f]
  /// \f$x \in \mathbb{R}^n\f$, \f$f(x) \in \mathbb{R}^m\f$ where
  /// \f$n\f$ is the input size and \f$m\f$ is the output size.
  ///
  /// Gradient computation is done through the #impl_gradient method
  /// that has to implemented by the concrete class inheriting this
  /// class.
  ///
  /// Jacobian computation is automatically done by concatenating
  /// gradients together, however this naive implementation can be
  /// overridden by the concrete class.
  ///
  ///
  /// The gradient of a \f$\mathbb{R}^n \rightarrow \mathbb{R}^m\f$
  /// function where \f$n > 1\f$ and \f$m > 1\f$ is a matrix.
  /// As this representation is costly, RobOptim considers
  /// these functions as \f$m\f$ \f$\mathbb{R}^n \rightarrow \mathbb{R}\f$
  /// functions. Through that mechanism, gradients are always vectors
  /// and jacobian are always matrices.
  /// When the gradient or the jacobian has to be computed, one has to
  /// precise which of the \f$m\f$ functions should be considered.
  ///
  /// If \f$m = 1\f$, then the function id must always be 0 and can be safely
  /// ignored in the gradient/jacobian computation.
  /// The class provides a default value for the function id so that these functions
  /// do not have to explicitly set the function id.
  class DerivableFunction : public Function
  {
  public:
    /// \brief Gradient type.
    typedef vector_t gradient_t;
    /// \brief Jacobian type.
    typedef matrix_t jacobian_t;

    /// \brief Jacobian size type (pair of values).
    typedef std::pair<value_type, value_type> jacobianSize_t;


    /// \brief Return the gradient size.
    ///
    /// Gradient size is equals to the input size.
    size_t gradientSize () const throw ()
    {
      return inputSize ();
    }

    /// \brief Return the jacobian size as a pair.
    ///
    /// Gradient size is equals to (output size, input size).
    jacobianSize_t jacobianSize () const throw ()
    {
      return std::make_pair (outputSize (), inputSize ());
    }

    /// \brief Check if the gradient is valid (check size).
    /// \param gradient checked gradient
    /// \return true if valid, false if not
    bool isValidGradient (const gradient_t& gradient) const throw ()
    {
      return gradient.size () == gradientSize ();
    }

    /// \brief Check if the jacobian is valid (check sizes).
    ///
    /// \param jacobian checked jacobian
    /// \return true if valid, false if not
    bool isValidJacobian (const jacobian_t& jacobian) const throw ()
    {
      return jacobian.size1 () == jacobianSize ().first
	&& jacobian.size2 () == jacobianSize ().second;
    }

    /// \brief Computes the jacobian.
    ///
    /// \param argument point at which the jacobian will be computed
    /// \return jacobian matrix
    jacobian_t jacobian (const argument_t& argument) const throw ()
    {
      jacobian_t jacobian (jacobianSize ().first, jacobianSize ().second);
      jacobian.clear ();
      this->jacobian (jacobian, argument);
      return jacobian;
    }

    /// \brief Computes the jacobian.
    ///
    /// Program will abort if the jacobian size is wrong before
    /// or after the jacobian computation.
    /// \param jacobian jacobian will be stored in this argument
    /// \param argument point at which the jacobian will be computed
    void jacobian (jacobian_t& jacobian, const argument_t& argument) const throw ()
    {
      assert (argument.size () == inputSize ());
      assert (isValidJacobian (jacobian));
      this->impl_jacobian (jacobian, argument);
      assert (isValidJacobian (jacobian));
    }

    /// \brief Computes the gradient.
    ///
    /// \param argument point at which the gradient will be computed
    /// \param functionId function id in split representation
    /// \return gradient vector
    gradient_t gradient (const argument_t& argument,
			 size_type functionId = 0) const throw ()
    {
      gradient_t gradient (gradientSize ());
      gradient.clear ();
      this->gradient (gradient, argument, functionId);
      return gradient;
    }

    /// \brief Computes the gradient.
    ///
    /// Program will abort if the gradient size is wrong before
    /// or after the gradient computation.
    /// \param gradient gradient will be stored in this argument
    /// \param argument point at which the gradient will be computed
    /// \param functionId function id in split representation
    /// \return gradient vector
    void gradient (gradient_t& gradient,
		   const argument_t& argument,
		   size_type functionId = 0) const throw ()
    {
      assert (argument.size () == inputSize ());
      assert (isValidGradient (gradient));
      this->impl_gradient (gradient, argument, functionId);
      assert (isValidGradient (gradient));
    }

    /// \brief Display the function on the specified output stream.
    ///
    /// \param o output stream used for display
    /// \return output stream
    virtual std::ostream& print (std::ostream& o) const throw ();

  protected:
    /// \brief Concrete class constructor should call this constructor.
    ///
    /// \param inputSize input size (argument size)
    /// \param outputSize output size (result size)
    DerivableFunction (size_type inputSize, size_type outputSize = 1) throw ();

    /// \brief Jacobian evaluation.
    ///
    /// Computes the jacobian, can be overridden by concrete classes.
    /// The default behavior is to compute the jacobian from the gradient.
    /// \warning Do not call this function directly, call #jacobian instead.
    /// \param jacobian jacobian will be store in this argument
    /// \param arg point where the jacobian will be computed
    virtual void impl_jacobian (jacobian_t& jacobian, const argument_t& arg)
      const throw ();

    /// \brief Gradient evaluation.
    ///
    /// Compute the gradient, has to be implemented in concrete classes.
    /// The gradient is computed for a specific sub-function which id
    /// is passed through the functionId argument.
    /// \warning Do not call this function directly, call #gradient instead.
    /// \param gradient gradient will be store in this argument
    /// \param argument point where the gradient will be computed
    /// \param functionId evaluated function id in the split representation
    virtual void impl_gradient (gradient_t& gradient,
				const argument_t& argument,
				size_type functionId = 0)
      const throw () = 0;
  };

  /// @}

} // end of namespace roboptim
#endif //! ROBOPTIM_CORE_DERIVABLE_FUNCTION_HH
