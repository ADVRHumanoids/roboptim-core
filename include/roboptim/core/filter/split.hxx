// Copyright (C) 2010 by Thomas Moulard, AIST, CNRS, INRIA.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

#ifndef ROBOPTIM_CORE_FILTER_SPLIT_HXX
# define ROBOPTIM_CORE_FILTER_SPLIT_HXX
# include <boost/format.hpp>

# include <roboptim/core/derivative-size.hh>

namespace roboptim
{
  namespace
  {
    template <typename T>
    std::string splitName (const T& fct);

    template <typename T>
    std::string splitName (const T& fct, unsigned functionId)
    {
      boost::format fmt ("%1% (split, function Id = %2%)");
      fmt % fct.getName () % functionId;
      return fmt.str ();
    }
  } // end of anonymous namespace.

  template <typename T>
  Split<T>::Split (boost::shared_ptr<const T> fct, unsigned functionId) throw ()
    : T (fct->inputSize (), 1, splitName (*fct, functionId)),
      function_ (fct),
      functionId_ (functionId)
  {
    assert (functionId < fct->outputSize ());
  }

  template <typename T>
  Split<T>::~Split () throw ()
  {
  }

  template <typename T>
  void
  Split<T>::impl_compute (result_t& result,
			  const argument_t& argument)
    const throw ()
  {
    result_t res (function_->outputSize ());
    (*function_) (res, argument);
    result[0] = res[functionId_];
  }


  template <>
  void
  Split<Function>::impl_gradient (gradient_t&, const argument_t&, size_type)
    const throw ()
  {
    assert (0);
  }

  template <typename T>
  void
  Split<T>::impl_gradient (gradient_t& gradient,
			   const argument_t& argument,
			   size_type functionId)
    const throw ()
  {
    assert (functionId == 0);
    function_->gradient (gradient, argument, functionId_);
  }




  template <>
  void
  Split<Function>::impl_hessian
  (hessian_t&, const argument_t&, size_type) const throw ()
  {
    assert (0);
  }

  template <>
  void
  Split<DerivableFunction>::impl_hessian
  (hessian_t&, const argument_t&, size_type) const throw ()
  {
    assert (0);
  }



  template <typename T>
  void
  Split<T>::impl_hessian (hessian_t& hessian,
			  const argument_t& argument,
			  size_type functionId)
    const throw ()
  {
    assert (functionId == 0);
    function_->hessian (hessian, argument, functionId_);
  }


  template <>
  void
  Split<Function>::impl_derivative
  (gradient_t&, double, size_type) const throw ()
  {
    assert (0);
  }

  template <>
  void
  Split<DerivableFunction>::impl_derivative
  (gradient_t&, double, size_type) const throw ()
  {
    assert (0);
  }

  template <>
  void
  Split<TwiceDerivableFunction>::impl_derivative
  (gradient_t&, double, size_type) const throw ()
  {
    assert (0);
  }

  template <typename T>
  void
  Split<T>::impl_derivative (gradient_t& derivative,
			     double argument,
			     size_type order)
    const throw ()
  {
    function_->derivative (derivative, argument, order);
  }

  template <typename P, typename C>
  void addNonScalarConstraint
  (P& problem,
   boost::shared_ptr<C> constraint,
   std::vector<Function::interval_t> interval,
   std::vector<Function::value_type> scale)
    throw (std::runtime_error)
  {
    assert (constraint);
    assert (interval.size () == constraint->outputSize ());
    assert (scale.empty () || scale.size () == constraint->outputSize ());

    if (constraint->outputSize () == 1)
    {
      if (scale.empty ())
	problem.addConstraint (constraint, interval[0]);
      else
	problem.addConstraint (constraint, interval[0], scale[0]);
      return;
    }
    for (unsigned i = 0; i < constraint->outputSize (); ++i)
      {
	boost::shared_ptr<Split<C> > split (new Split<C> (constraint, i));
	if (scale.empty ())
	  problem.addConstraint (split, interval[i]);
	else
	  problem.addConstraint (split, interval[i], scale[i]);
      }
  }
} // end of namespace roboptim

#endif //! ROBOPTIM_CORE_FILTER_SPLIT_HXX
