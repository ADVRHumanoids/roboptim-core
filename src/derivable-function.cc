// Copyright (C) 2009 by Thomas Moulard, FIXME.
//
// This file is part of the liboptimization.
//
// liboptimization is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// liboptimization is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with liboptimization.  If not, see <http://www.gnu.org/licenses/>.


/**
 * \brief Implementation of the DerivableFunction class.
 */

#include "liboptimization/derivable-function.hh"

namespace optimization
{
  DerivableFunction::DerivableFunction (size_type n, value_type infinity)
    throw ()
    : Function (n, infinity)
  {
  }

} // end of namespace optimization
