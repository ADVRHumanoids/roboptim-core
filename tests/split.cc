// Copyright (C) 2010 by Thomas Moulard, AIST, CNRS, INRIA.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

#undef NDEBUG

#include "shared-tests/common.hh"

#include <iostream>

#include <roboptim/core/io.hh>
#include <roboptim/core/derivable-function.hh>
#include <roboptim/core/util.hh>
#include <roboptim/core/filter/split.hh>

using namespace roboptim;

struct F : public DerivableFunction
{
  F () : DerivableFunction (1, 10, "f_n (x) = n * x")
  {}

  void impl_compute (result_t& res, const argument_t& argument) const throw ()
  {
    res.clear ();
    for (size_type i = 0; i < outputSize (); ++i)
      res[i] = i * argument[0];
  }

  void impl_gradient (gradient_t& grad, const argument_t&,
		      size_type functionId) const throw ()
  {
    grad.clear ();
    grad[0] = functionId;
  }
};

int run_test ()
{
  boost::shared_ptr<F> f (new F ());

  for (unsigned id = 0; id < 10; ++id)
    {
      Split<DerivableFunction> splitF (f, id);

      std::cout << splitF << ":" << std::endl
		<< std::endl;
      
      Function::vector_t x (1);
      for (double i = 0.; i < 10.; i += 0.5)
	{
	  x[0] = i;
	  std::cout << splitF (x) << std::endl;
	  std::cout << splitF (x) << std::endl;
	  //assert ((*f) (x)[0] == splitF (x)[id]);
	  
	  std::cout << splitF.gradient (x) << std::endl;
	  std::cout << splitF.gradient (x) << std::endl;
	}
    }
  return 0;
}

GENERATE_TEST ()
