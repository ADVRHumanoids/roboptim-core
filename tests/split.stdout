f_n (x) = n * x (split, function Id = 0) (differentiable function):

[1](0)
[1](0)
[1](0)
[1](0)
[1](0)
[1](0)
[1](0)
[1](0)
[1](0)
[1](0)
[1](0)
[1](0)
[1](0)
[1](0)
[1](0)
[1](0)
[1](0)
[1](0)
[1](0)
[1](0)
[1](0)
[1](0)
[1](0)
[1](0)
[1](0)
[1](0)
[1](0)
[1](0)
[1](0)
[1](0)
[1](0)
[1](0)
[1](0)
[1](0)
[1](0)
[1](0)
[1](0)
[1](0)
[1](0)
[1](0)
[1](0)
[1](0)
[1](0)
[1](0)
[1](0)
[1](0)
[1](0)
[1](0)
[1](0)
[1](0)
[1](0)
[1](0)
[1](0)
[1](0)
[1](0)
[1](0)
[1](0)
[1](0)
[1](0)
[1](0)
[1](0)
[1](0)
[1](0)
[1](0)
[1](0)
[1](0)
[1](0)
[1](0)
[1](0)
[1](0)
[1](0)
[1](0)
[1](0)
[1](0)
[1](0)
[1](0)
[1](0)
[1](0)
[1](0)
[1](0)
f_n (x) = n * x (split, function Id = 1) (differentiable function):

[1](0)
[1](0)
[1](1)
[1](1)
[1](0.5)
[1](0.5)
[1](1)
[1](1)
[1](1)
[1](1)
[1](1)
[1](1)
[1](1.5)
[1](1.5)
[1](1)
[1](1)
[1](2)
[1](2)
[1](1)
[1](1)
[1](2.5)
[1](2.5)
[1](1)
[1](1)
[1](3)
[1](3)
[1](1)
[1](1)
[1](3.5)
[1](3.5)
[1](1)
[1](1)
[1](4)
[1](4)
[1](1)
[1](1)
[1](4.5)
[1](4.5)
[1](1)
[1](1)
[1](5)
[1](5)
[1](1)
[1](1)
[1](5.5)
[1](5.5)
[1](1)
[1](1)
[1](6)
[1](6)
[1](1)
[1](1)
[1](6.5)
[1](6.5)
[1](1)
[1](1)
[1](7)
[1](7)
[1](1)
[1](1)
[1](7.5)
[1](7.5)
[1](1)
[1](1)
[1](8)
[1](8)
[1](1)
[1](1)
[1](8.5)
[1](8.5)
[1](1)
[1](1)
[1](9)
[1](9)
[1](1)
[1](1)
[1](9.5)
[1](9.5)
[1](1)
[1](1)
f_n (x) = n * x (split, function Id = 2) (differentiable function):

[1](0)
[1](0)
[1](2)
[1](2)
[1](1)
[1](1)
[1](2)
[1](2)
[1](2)
[1](2)
[1](2)
[1](2)
[1](3)
[1](3)
[1](2)
[1](2)
[1](4)
[1](4)
[1](2)
[1](2)
[1](5)
[1](5)
[1](2)
[1](2)
[1](6)
[1](6)
[1](2)
[1](2)
[1](7)
[1](7)
[1](2)
[1](2)
[1](8)
[1](8)
[1](2)
[1](2)
[1](9)
[1](9)
[1](2)
[1](2)
[1](10)
[1](10)
[1](2)
[1](2)
[1](11)
[1](11)
[1](2)
[1](2)
[1](12)
[1](12)
[1](2)
[1](2)
[1](13)
[1](13)
[1](2)
[1](2)
[1](14)
[1](14)
[1](2)
[1](2)
[1](15)
[1](15)
[1](2)
[1](2)
[1](16)
[1](16)
[1](2)
[1](2)
[1](17)
[1](17)
[1](2)
[1](2)
[1](18)
[1](18)
[1](2)
[1](2)
[1](19)
[1](19)
[1](2)
[1](2)
f_n (x) = n * x (split, function Id = 3) (differentiable function):

[1](0)
[1](0)
[1](3)
[1](3)
[1](1.5)
[1](1.5)
[1](3)
[1](3)
[1](3)
[1](3)
[1](3)
[1](3)
[1](4.5)
[1](4.5)
[1](3)
[1](3)
[1](6)
[1](6)
[1](3)
[1](3)
[1](7.5)
[1](7.5)
[1](3)
[1](3)
[1](9)
[1](9)
[1](3)
[1](3)
[1](10.5)
[1](10.5)
[1](3)
[1](3)
[1](12)
[1](12)
[1](3)
[1](3)
[1](13.5)
[1](13.5)
[1](3)
[1](3)
[1](15)
[1](15)
[1](3)
[1](3)
[1](16.5)
[1](16.5)
[1](3)
[1](3)
[1](18)
[1](18)
[1](3)
[1](3)
[1](19.5)
[1](19.5)
[1](3)
[1](3)
[1](21)
[1](21)
[1](3)
[1](3)
[1](22.5)
[1](22.5)
[1](3)
[1](3)
[1](24)
[1](24)
[1](3)
[1](3)
[1](25.5)
[1](25.5)
[1](3)
[1](3)
[1](27)
[1](27)
[1](3)
[1](3)
[1](28.5)
[1](28.5)
[1](3)
[1](3)
f_n (x) = n * x (split, function Id = 4) (differentiable function):

[1](0)
[1](0)
[1](4)
[1](4)
[1](2)
[1](2)
[1](4)
[1](4)
[1](4)
[1](4)
[1](4)
[1](4)
[1](6)
[1](6)
[1](4)
[1](4)
[1](8)
[1](8)
[1](4)
[1](4)
[1](10)
[1](10)
[1](4)
[1](4)
[1](12)
[1](12)
[1](4)
[1](4)
[1](14)
[1](14)
[1](4)
[1](4)
[1](16)
[1](16)
[1](4)
[1](4)
[1](18)
[1](18)
[1](4)
[1](4)
[1](20)
[1](20)
[1](4)
[1](4)
[1](22)
[1](22)
[1](4)
[1](4)
[1](24)
[1](24)
[1](4)
[1](4)
[1](26)
[1](26)
[1](4)
[1](4)
[1](28)
[1](28)
[1](4)
[1](4)
[1](30)
[1](30)
[1](4)
[1](4)
[1](32)
[1](32)
[1](4)
[1](4)
[1](34)
[1](34)
[1](4)
[1](4)
[1](36)
[1](36)
[1](4)
[1](4)
[1](38)
[1](38)
[1](4)
[1](4)
f_n (x) = n * x (split, function Id = 5) (differentiable function):

[1](0)
[1](0)
[1](5)
[1](5)
[1](2.5)
[1](2.5)
[1](5)
[1](5)
[1](5)
[1](5)
[1](5)
[1](5)
[1](7.5)
[1](7.5)
[1](5)
[1](5)
[1](10)
[1](10)
[1](5)
[1](5)
[1](12.5)
[1](12.5)
[1](5)
[1](5)
[1](15)
[1](15)
[1](5)
[1](5)
[1](17.5)
[1](17.5)
[1](5)
[1](5)
[1](20)
[1](20)
[1](5)
[1](5)
[1](22.5)
[1](22.5)
[1](5)
[1](5)
[1](25)
[1](25)
[1](5)
[1](5)
[1](27.5)
[1](27.5)
[1](5)
[1](5)
[1](30)
[1](30)
[1](5)
[1](5)
[1](32.5)
[1](32.5)
[1](5)
[1](5)
[1](35)
[1](35)
[1](5)
[1](5)
[1](37.5)
[1](37.5)
[1](5)
[1](5)
[1](40)
[1](40)
[1](5)
[1](5)
[1](42.5)
[1](42.5)
[1](5)
[1](5)
[1](45)
[1](45)
[1](5)
[1](5)
[1](47.5)
[1](47.5)
[1](5)
[1](5)
f_n (x) = n * x (split, function Id = 6) (differentiable function):

[1](0)
[1](0)
[1](6)
[1](6)
[1](3)
[1](3)
[1](6)
[1](6)
[1](6)
[1](6)
[1](6)
[1](6)
[1](9)
[1](9)
[1](6)
[1](6)
[1](12)
[1](12)
[1](6)
[1](6)
[1](15)
[1](15)
[1](6)
[1](6)
[1](18)
[1](18)
[1](6)
[1](6)
[1](21)
[1](21)
[1](6)
[1](6)
[1](24)
[1](24)
[1](6)
[1](6)
[1](27)
[1](27)
[1](6)
[1](6)
[1](30)
[1](30)
[1](6)
[1](6)
[1](33)
[1](33)
[1](6)
[1](6)
[1](36)
[1](36)
[1](6)
[1](6)
[1](39)
[1](39)
[1](6)
[1](6)
[1](42)
[1](42)
[1](6)
[1](6)
[1](45)
[1](45)
[1](6)
[1](6)
[1](48)
[1](48)
[1](6)
[1](6)
[1](51)
[1](51)
[1](6)
[1](6)
[1](54)
[1](54)
[1](6)
[1](6)
[1](57)
[1](57)
[1](6)
[1](6)
f_n (x) = n * x (split, function Id = 7) (differentiable function):

[1](0)
[1](0)
[1](7)
[1](7)
[1](3.5)
[1](3.5)
[1](7)
[1](7)
[1](7)
[1](7)
[1](7)
[1](7)
[1](10.5)
[1](10.5)
[1](7)
[1](7)
[1](14)
[1](14)
[1](7)
[1](7)
[1](17.5)
[1](17.5)
[1](7)
[1](7)
[1](21)
[1](21)
[1](7)
[1](7)
[1](24.5)
[1](24.5)
[1](7)
[1](7)
[1](28)
[1](28)
[1](7)
[1](7)
[1](31.5)
[1](31.5)
[1](7)
[1](7)
[1](35)
[1](35)
[1](7)
[1](7)
[1](38.5)
[1](38.5)
[1](7)
[1](7)
[1](42)
[1](42)
[1](7)
[1](7)
[1](45.5)
[1](45.5)
[1](7)
[1](7)
[1](49)
[1](49)
[1](7)
[1](7)
[1](52.5)
[1](52.5)
[1](7)
[1](7)
[1](56)
[1](56)
[1](7)
[1](7)
[1](59.5)
[1](59.5)
[1](7)
[1](7)
[1](63)
[1](63)
[1](7)
[1](7)
[1](66.5)
[1](66.5)
[1](7)
[1](7)
f_n (x) = n * x (split, function Id = 8) (differentiable function):

[1](0)
[1](0)
[1](8)
[1](8)
[1](4)
[1](4)
[1](8)
[1](8)
[1](8)
[1](8)
[1](8)
[1](8)
[1](12)
[1](12)
[1](8)
[1](8)
[1](16)
[1](16)
[1](8)
[1](8)
[1](20)
[1](20)
[1](8)
[1](8)
[1](24)
[1](24)
[1](8)
[1](8)
[1](28)
[1](28)
[1](8)
[1](8)
[1](32)
[1](32)
[1](8)
[1](8)
[1](36)
[1](36)
[1](8)
[1](8)
[1](40)
[1](40)
[1](8)
[1](8)
[1](44)
[1](44)
[1](8)
[1](8)
[1](48)
[1](48)
[1](8)
[1](8)
[1](52)
[1](52)
[1](8)
[1](8)
[1](56)
[1](56)
[1](8)
[1](8)
[1](60)
[1](60)
[1](8)
[1](8)
[1](64)
[1](64)
[1](8)
[1](8)
[1](68)
[1](68)
[1](8)
[1](8)
[1](72)
[1](72)
[1](8)
[1](8)
[1](76)
[1](76)
[1](8)
[1](8)
f_n (x) = n * x (split, function Id = 9) (differentiable function):

[1](0)
[1](0)
[1](9)
[1](9)
[1](4.5)
[1](4.5)
[1](9)
[1](9)
[1](9)
[1](9)
[1](9)
[1](9)
[1](13.5)
[1](13.5)
[1](9)
[1](9)
[1](18)
[1](18)
[1](9)
[1](9)
[1](22.5)
[1](22.5)
[1](9)
[1](9)
[1](27)
[1](27)
[1](9)
[1](9)
[1](31.5)
[1](31.5)
[1](9)
[1](9)
[1](36)
[1](36)
[1](9)
[1](9)
[1](40.5)
[1](40.5)
[1](9)
[1](9)
[1](45)
[1](45)
[1](9)
[1](9)
[1](49.5)
[1](49.5)
[1](9)
[1](9)
[1](54)
[1](54)
[1](9)
[1](9)
[1](58.5)
[1](58.5)
[1](9)
[1](9)
[1](63)
[1](63)
[1](9)
[1](9)
[1](67.5)
[1](67.5)
[1](9)
[1](9)
[1](72)
[1](72)
[1](9)
[1](9)
[1](76.5)
[1](76.5)
[1](9)
[1](9)
[1](81)
[1](81)
[1](9)
[1](9)
[1](85.5)
[1](85.5)
[1](9)
[1](9)
